qiskit/providers/aer/backends/aerbackend.py

```diff
@@ -330,11 +331,21 @@ def _format_results(output):
                     data[key] = format_save_type(val, save_types[key], save_subtypes[key])
         return Result.from_dict(output)
 
+    @staticmethod
+    def _circuit_optypes(circuit):
+        """Retur set of all operation names in a circuit"""
+        if not isinstance(circuit, QuantumCircuit):
+            return set()
+        optypes = set()
+        for instr, _, _ in circuit._data:
+            optypes.add(type(instr))
```

Comment:
I think you need to recurse into control-flow ops here.  It'll work for the use case you have in mind, because the only way you could be "hiding" a control-flow op is if it's inside another (so it'd be detected), but it's basically a latent bug waiting to happen otherwise.

Reply:
This is a good point. I suppose a more robust thing should be to do the control flow unrolling first, before the noise assembling, and you rebuild the opset after doing each unrolling.